### ISSUE TITLE
Allow adding extra parameters to body/header in Oauth2 flow

### ISSUE THREAD

**Issue by mdomashchenko** [a85/POSTMan-Chrome-Extension#2523]
1. Postman Version: v4.9.1 
 2. App (Chrome app or Mac app): Standalone (Electron)
 3. OS details: Windows 10
 4. Is the Interceptor on and enabled in the app:
 5. Did you encounter this recently, or has this bug always been there: I've tried this today for the first time, encountered the bug right away.
 6. Expected behaviour: Should give me an option to send resource parameter with authorization_code request to oauth2/token endpoint.

I was trying to use Postman to make API requests to https://management.azure.com/. This requires OAuth 2.0 token, so I clicked "Get New Access Token", used https://login.microsoftonline.com/{my tenant}/oauth2/authorize for Auth URL, https://login.microsoftonline.com/{my tenant}/oauth2/token for Auth Token URL, no scope, Authorization Code grant type, Request locally checked.

Click "Request Token" nothing much happens. Started debugging with Fiddler. Everything goes as it should but the last authorization_code request to the oauth2/token returns error "AADSTS50001: Resource identifier is not provided". The request included client_id, redirect_url (getpostman callback, as requested), grant_type: authorization_code, state and code.

I set a "before request" breakpoint in Fiddler, catch that last request and add parameter "resource" with value "https://management.azure.com/". This returns correct response allowing Postman to obtain the bearer token that I could use successfully to make Azure REST API requests.

It would be really nice if the "Get New Access Token" dialog included an input for the resource, to avoid the hassle with Fiddler breakpoiunts.

**Comment by rgregg** [comment]
+1. I'm hitting this issue as well.

**Comment by jongio** [comment]
+1. In the meantime, I got it working with this approach: http://blog.jongallant.com/2017/03/azure-active-directory-access-tokens-postman/

I also have a posts to help generate SAS Tokens in the Pre-request Script sandbox: http://blog.jongallant.com/tags/postman/

**Comment by kamoljan** [comment]
+1

**Comment by dennisesan** [comment]
+1

**Comment by harryi3t** [comment]
@mike69d We are fixing this bug. But for now, I can suggest you a workaround.
 Can you try __un-checking__ the `Request access token locally` and let us know if that works?

 <img src= 'https://user-images.githubusercontent.com/5207331/29372418-aa89146a-82c8-11e7-9965-75dbdda990c0.png' height=500px/>

Marking this as possible duplicate of https://github.com/postmanlabs/postman-app-support/issues/2296

**Comment by RodrigoGroener** [comment]
+1

**Comment by pipe01** [comment]
+1

**Comment by harryi3t** [comment]
@mike69d @pipe01 @RodrigoGroener 
Just to make sure I understand this issue correctly, the `resource` parameter which is missing is not actually part of the RFC?

According to the [rfc#6749 section-4.1.3](https://tools.ietf.org/html/rfc6749#section-4.1.3) these are the parameters that the server requires in the token endpoint

> grant_type
>          REQUIRED.  Value MUST be set to "authorization_code".
> 
>    code
>          REQUIRED.  The authorization code received from the
>          authorization server.
> 
>    redirect_uri
>          REQUIRED, if the "redirect_uri" parameter was included in the
>          authorization request as described in Section 4.1.1, and their
>          values MUST be identical.
> 
>    client_id
>          REQUIRED, if the client is not authenticating with the
>          authorization server as described in Section 3.2.1.

If this is true then it's a feature request and a not a bug, right?
Please correct me if I am wrong

**Comment by salanford** [comment]
I was butting heads with this too and discovered that the resource value can be passed in the query string instead of the body. So your authorize endpoint would look something like this: https://login.microsoftonline.com/{{tenantId}}/oauth2/authorize?resource={{resource}}

**Comment by harryi3t** [comment]
@salanford Thank you very much for the workaround.
Closing this now since there is nothing else required here.

**Comment by renelc30** [comment]
@salanford thanks so much for posting the trick. Did it for me

**Comment by jongio** [comment]
This doesn't work with the client_credentials flow.
![image](https://user-images.githubusercontent.com/2163001/33677721-8ac9bc44-da6e-11e7-97c0-a86067dd64e9.png)


@postman-admin Can you add the ability to specify additional body parameters so I can add a "resource" parameter to satisfy Azure's resource requirement.  I'm asking to add a generic way to add parameters, not Azure specific.  Thanks.

**Comment by Enver-Yilmaz** [comment]
+1 for suggestion from @jonbgallant

**Comment by andreipet** [comment]
+1 for suggestion from @salanford 
Optional parameter....but without it you will get AADSTS50001...funny

**Comment by gs11** [comment]
@harryi3t Please reopen this. See comment from @jongio on the issue with the client credentials flow.

**Comment by harryi3t** [comment]
@gs11 Reopening. I am marking this as a feature instead of a bug.

**Comment by tiloc** [comment]
+1 On my deployment of Azure I need to add the "Ocp-Apim-Subscription-Key" to the transmitted headers when requesting an OAuth2.0 token. Need the possibility to add this to the token request.

**Comment by tychota** [comment]
For people needing to add a `X-Client` header (see #4774), you can work around in between using for instance Charles Proxy (https://www.charlesproxy.com/download/, which is not free but has a demo version)

![image](https://user-images.githubusercontent.com/13785185/42314592-a8b8d078-8045-11e8-9e9f-b6ffc104bd2e.png)

I disabled the Mac OSx proxy not to pollute the navigator by adding / overriding headers that may already exists.

![image](https://user-images.githubusercontent.com/13785185/42314680-db55aa56-8045-11e8-9e17-779b0f8ec3b3.png)

I activated ssl proxyfing

![image](https://user-images.githubusercontent.com/13785185/42314706-e8dd244c-8045-11e8-8bb6-574a4f6db025.png)

Then in postman I disabled ssl verification

![image](https://user-images.githubusercontent.com/13785185/42314777-16c19212-8046-11e8-9889-10c7b534bae4.png)

And activated the proxy:

![image](https://user-images.githubusercontent.com/13785185/42314841-37bfcc9a-8046-11e8-95f7-3c9bb9d0f77c.png)

Then you can use the oauth2 flow properly.

**Comment by jongio** [comment]
@numaanashraf - Do you know if this is on a backlog or have an ETA?

**Comment by 123dev** [comment]
Can't believe 4 years later and this is still in Triage.

**Comment by vikramksrm** [comment]
+1

**Comment by Yvand** [comment]
I face the same issue with grant type "Authorization code", in scenario described in [this article](https://docs.microsoft.com/en-us/sharepoint/dev/sp-add-ins/authorization-code-oauth-flow-for-sharepoint-add-ins).
I found no workaround because parameter "resource" must be sent in the body, I tested that it is ignored if sent in the query string.

Error message returned by ACS from the POST to https://accounts.accesscontrol.windows.net/{{TenantId}}/tokens/OAuth/2:
```
error_description=AADSTS900144: The request body must contain the following parameter: 'resource'.
Trace ID: 0451e9fc-149b-42f5-a088-a6328ce61b00
Correlation ID: f42c3d44-8a78-4249-9c11-f2cffb6b1365
Timestamp: 2020-12-18 13:52:26Z
```

Can you guys give visibility if/when you are going to make possible to add custom parameters in the body?

**Comment by bneumann** [comment]
For what it's worth I stumbled across this issue (with Azure OAuth, which wants the resource key) and couldn't use the build in functionality. I made a script that runs before each query that sets a variable which then again can be consumed by the query:

```js
const formdata =[
    { key: "client_id", value: "your_client_id" },
    { key: "client_secret",  value: "your_client_secret" },
    { key: "grant_type",  value: "client_credentials" },
    { key: "resource",  value: "your_uuid" }
];

const url = "your_oauth_url_here";

pm.sendRequest({
        url: url,
        method: "POST",
        body:
        {
            mode: "formdata",
            formdata: formdata
        }
    },
    function(err, response) {
        if(err){
            console.log(err);
            return;
        }
        console.log("Token aquired");
        pm.collectionVariables.set("token", response.json().access_token);
    }
);
```
I use the token in the same folder then:

![image](https://user-images.githubusercontent.com/2605131/113114327-82b30d00-920b-11eb-8f48-d24b22209d28.png)

This is not the proper solution for extending the header/body but in my case a proper workaround. Plus I won't need to update the token at all, it is always fetched (yeah I know it's not good style :) ).

**Comment by jongio** [comment]
Azure REST APIs with Postman using Pre-request Script that handles token expiration
- Blog: https://aka.ms/azurerestblog
- Video: https://aka.ms/azurerestvideo

Azure REST APIs with Postman's OAuth 2.0 Provider: 
- Blog: https://blog.jongallant.com/2021/03/azure-rest-apis-postman-oauth2/

**Comment by giridharvc7** [comment]
The additional options for OAuth2 , "audience" and "resource" are now available on [Postman Web](https://www.postman.com/home) . Please give it a try, and let us know if you have any feedback or face any issues. This feature will be available in the desktop apps with the 8.3 release
![image](https://user-images.githubusercontent.com/1040583/115075056-2264e080-9f18-11eb-91a4-31a37dab2a10.png)

**Comment by giridharvc7** [comment]
Can you please try this and let us know if this solves @jongio @bneumann @123dev ? 
Closing this issue for now.

**Comment by jongio** [comment]
Thanks you!

**Comment by eric-gagnon** [comment]
> The additional options for OAuth2 , "audience" and "resource" are now available on [Postman Web](https://www.postman.com/home) . Please give it a try, and let us know if you have any feedback or face any issues. This feature will be available in the desktop apps with the 8.3 release ![image](https://user-images.githubusercontent.com/1040583/115075056-2264e080-9f18-11eb-91a4-31a37dab2a10.png)

It looks like there is a little bug, it was present in 8.x but I still see it after updating to lasted 9.x. The "resource" option was not visible. 

![image](https://user-images.githubusercontent.com/21067311/184988284-aed1f0d1-2330-4bb1-b6c1-1e554d12d187.png)

By luck, I tought to check if it was exposed in other grant type and it was. So I set the value, changed back my grant type to client credential and now it's visible in advanced options.

![image](https://user-images.githubusercontent.com/21067311/184988231-df845ce8-2d22-4d76-927d-55781acbe867.png)
